src/Compilers/CSharp/Test/Emit/PDB/CSharpPDBTestBase.cs

```diff
@@ -17,6 +20,64 @@ namespace Microsoft.CodeAnalysis.CSharp.UnitTests.PDB
+        {
+            MarkupTestFile.GetSpans(input, out var source, out IDictionary<string, ImmutableArray<TextSpan>> spans);
+
+            var compilation = CreateCompilationWithMscorlib40AndSystemCore(source, options: TestOptions.DebugDll);
+            compilation.GetDiagnostics().Where(d => d.Severity == DiagnosticSeverity.Error).Verify();
+
+            var pdb = PdbValidation.GetPdbXml(compilation);
+
+            //TODO: Should we support having more than one method under test? Or does that make for tests that do too much?
+            var methodStart = spans[methodStartSpanName].Single().Start;
+
+            // We need to offset the closures and lambdas by the method start point
+            var closures = spans[closureSpanName].SelectAsArray(c => new TextSpan(c.Start - methodStart, c.Length));
+            var lambdas = spans[lambdaSpanName].SelectAsArray(c => new TextSpan(c.Start - methodStart, c.Length));
+
+            CheckClosuresAndLambdas(pdb, closures, lambdas);
+        }
+
+        private static void CheckClosuresAndLambdas(string pdb, ImmutableArray<TextSpan> closures, ImmutableArray<TextSpan> lambdas)
+        {
+            var pdbXml = XElement.Parse(pdb);
+
+            // Compare as strings so the developer gets nicer failure messages
+            var actualClosures = pdbXml.DescendantsAndSelf("closure").Select(c => c.Attribute("offset").Value).Select(int.Parse);
+
+            // Verify number and start position of closures
+            var actualClosuresDescription = GetOffsetsDescription("Closure", actualClosures);
+            var expectedClosuresDescription = GetOffsetsDescription("Closure", closures.Select(c => c.Start));
+            Assert.Equal(expectedClosuresDescription, actualClosuresDescription);
+
+            var actualLambdas = pdbXml.DescendantsAndSelf("lambda").Select(c => new
+            {
+                Offset = int.Parse(c.Attribute("offset").Value),
+                Closure = int.Parse(c.Attribute("closure").Value)
+            });
+
+            // Verify number and start position of lambdas
+            var actualLambdasDescription = GetOffsetsDescription("Lambda", actualLambdas.Select(l => l.Offset));
+            var expectedLambdasDescription = GetOffsetsDescription("Lambda", lambdas.Select(l => l.Start));
+            Assert.Equal(actualLambdasDescription, expectedLambdasDescription);
+
+            // Verify that lambdas are inside their expected closures
+
+            // The markup system has these ordered by span end, but PDB xml indexes by start
+            var orderedClosures = closures.OrderBy(c => c.Start).ToArray();
+            foreach (var lambda in actualLambdas)
+            {
+                var closure = orderedClosures[lambda.Closure];
+                Assert.True(closure.Contains(lambda.Offset), $"Lambda at offset {lambda.Offset} was expected to be contained in closure {lambda.Closure} but that closure is {closure}");
```

Comment:
>Assert.True(closure.Contains(lam [](start = 16, length = 32)

If the test fails we should print out the source with the markup, to make it easier to see what's expected.

Reply:
One of the downsides to this approach is that its not _really_ possible to output the expected source, because there is no information about where the closures (or lambdas) should end. The best it could do would be to put the closure end right after the last lambda that is contained within it, but I think that would be quite ugly.

Reply:
Yeah, that's a problem. We need to make this work though. Let's change the markers so that it's possible to print out the expected marked source.

We could use the approach that active statement tests use: `<Closure:{N}>` and `<Lambda:{M}.{N}>`, where _M_, _N_ are ordinals.